### ISSUE TITLE
Rule Proposal: object-curly-newline

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#6072]
From [requirePaddingNewLinesInObjects] and [disallowPaddingNewLinesInObjects].

This will require/disallow line breaks after open object curlies and before close object curlies.

## Options

```js
{
    "object-curly-newline": [
        "error", 
        "always" or "never" or {"multiline": <boolean>, "minItems": <integer>}
    ]
}
```

- `"always"` - Requires line breaks always.
- `"never"` - Disallows line breaks always.
- An object - Requires line breaks if any of properties is satisfied. Otherwise, disallows line breaks.
  - `multiline` (default is `false`) - Requires line breaks if the content is multiline. If this is `false`, this condition is disabled.
  - `minItems` (default is `0`) - Requires line breaks if the number of properties is more than the given integer. If this is `0`, this condition is disabled.

## Examples

```js
/*eslint object-curly-newline: ["error", "always"]*/

// ✔ GOOD
let a = {
};
let b = {
    foo: 1
};
let c = {
    foo: 1, bar: 2
};
let d = {
    foo: 1,
    bar: 2
};
let e = {
    foo: function() {
        dosomething();
    }
};

// ✘ BAD
let a = {};
let b = {foo: 1};
let c = {foo: 1, bar: 2};
let d = {foo: 1,
    bar: 2};
let e = {foo() {
    dosomething();
}};
```

```js
/*eslint object-curly-newline: ["error", "never"]*/

// ✔ GOOD
let a = {};
let b = {foo: 1};
let c = {foo: 1, bar: 2};
let d = {foo: 1,
    bar: 2};
let e = {foo: function() {
    dosomething();
}};

// ✘ BAD
let a = {
};
let b = {
    foo: 1
};
let c = {
    foo: 1, bar: 2
};
let d = {
    foo: 1,
    bar: 2
};
let e = {
    foo: function() {
        dosomething();
    }
};
```

```js
/*eslint object-curly-newline: ["error", {"multiline": true}]*/

// ✔ GOOD
let a = {};
let b = {foo: 1};
let c = {foo: 1, bar: 2};
let d = {
    foo: 1,
    bar: 2
};
let e = {
    foo: function() {
        dosomething();
    }
};

// ✘ BAD
let a = {
};
let b = {
    foo: 1
};
let c = {
    foo: 1, bar: 2
};
let d = {foo: 1,
    bar: 2};
let e = {foo: function() {
    dosomething();
}};
```

```js
/*eslint object-curly-newline: ["error", {"minItems": 2}]*/

// ✔ GOOD
let a = {};
let b = {foo: 1};
let c = {
    foo: 1, bar: 2
};
let d = {
    foo: 1,
    bar: 2
};
let e = {foo: function() {
    dosomething();
}};

// ✘ BAD
let a = {
};
let b = {
    foo: 1
};
let c = {foo: 1, bar: 2};
let d = {foo: 1,
    bar: 2};
let e = {
    foo: function() {
        dosomething();
    }
};
```

```js
/*eslint object-curly-newline: ["error", {"multiline": true "minItems": 2}]*/

// ✔ GOOD
let a = {};
let b = {foo: 1};
let c = {
    foo: 1, bar: 2
};
let d = {
    foo: 1,
    bar: 2
};
let e = {
    foo: function() {
        dosomething();
    }
};

// ✘ BAD
let a = {
};
let b = {
    foo: 1
};
let c = {foo: 1, bar: 2};
let d = {foo: 1,
    bar: 2};
let e = {foo: function() {
    dosomething();
}};
```

## Related Rules

|            | brace style                   | space style             | line break style              |
|-----------:|:------------------------------|:------------------------|:------------------------------|
| **block**  | [brace-style]                 | [block-spacing]         | [max-statements-per-line]     |
| **object** | [#6072 object-curly-newline]  | [object-curly-spacing]  | [object-property-newline]     |
| **array**  | [#6073 array-bracket-newline] | [array-bracket-spacing] | [#6075 array-element-newline] |

[brace-style]: http://eslint.org/docs/rules/brace-style
[block-spacing]: http://eslint.org/docs/rules/block-spacing
[max-statements-per-line]: http://eslint.org/docs/rules/max-statements-per-line
[object-curly-spacing]: http://eslint.org/docs/rules/object-curly-spacing
[object-property-newline]: http://eslint.org/docs/rules/object-property-newline
[array-bracket-spacing]: http://eslint.org/docs/rules/array-bracket-spacing
[#6072 object-curly-newline]: https://github.com/eslint/eslint/issues/6072
[#6073 array-bracket-newline]: https://github.com/eslint/eslint/issues/6073
[#6075 array-element-newline]: https://github.com/eslint/eslint/issues/6075

[requirePaddingNewLinesInObjects]: http://jscs.info/rule/requirePaddingNewLinesInObjects
[disallowPaddingNewLinesInObjects]: http://jscs.info/rule/disallowPaddingNewLinesInObjects

**Comment by ilyavolodin** [comment]
I think we already have an issue opened for something like that, having trouble finding it now. I think proposal was to add option to `brace-style` for object literals.

**Comment by mysticatea** [comment]
For spacing, rules are separated each kind of braces. So my proposals are, too.

`object-curly-spacing` ⇔ `newline-object-curly`
`array-bracket-spacing` ⇔ `newline-array-bracket`
`block-spacing` ⇔ `brace-style`

Indeed, `brace-style` rule is doing more things than spacing rules, so my proposals may not be enough.

**Comment by mysticatea** [comment]
@ilyavolodin It seems to be closed already: #5203

**Comment by mysticatea** [comment]
I updated this proposal.
If there is not objection opinions, I want to implement this.

**Comment by ilyavolodin** [comment]
:+1:

**Comment by nzakas** [comment]
@eslint/eslint-team need some more feedback here. 

👍

**Comment by BYK** [comment]
👍

**Comment by kaicataldo** [comment]
Unless there's already a convention to use items when referring to object props, I wonder if there is a better name for the minItems option. minProps/minProperties? A little more verbose, but that's the terminology I'm used to using/seeing.

Otherwise, :+1: from me

**Comment by nzakas** [comment]
@kaicataldo agreed, we should use "props" or "properties" for consistency

**Comment by mysticatea** [comment]
Thank you!

I'll update my PR by feedback.

**Comment by PhiLhoSoft** [comment]
#5203 isn't actually covered by this proposal.
For this, we would need a rule to "require/disallow line breaks BEFORE open object curlies" (and brackets too, in the other proposals).
Not sure if this must be included in this rule or addressed by a different rule.

**Comment by nzakas** [comment]
@PhiLhoSoft the goal here is not to implement a solution for #5203, it's to create an ESLint rule that is equivalent to two JSCS rules. Those rules only deal with inserting newlines inside of braces and that's what this issue is implementing.

**Comment by PhiLhoSoft** [comment]
Yes, I was answering the comment https://github.com/eslint/eslint/issues/6072#issuecomment-219219579 actually, referring to the above issue.
